File: app/src/main/java/com/kunzisoft/keepass/database/element/PwDatabase.java
/*
 * Copyright 2017 Brian Pellin, Jeremy Jamet / Kunzisoft.
 *     
 * This file is part of KeePass DX.
 *
 *  KeePass DX is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  KeePass DX is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with KeePass DX.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.kunzisoft.keepass.database.element;

import android.support.annotation.Nullable;

import com.kunzisoft.keepass.database.exception.InvalidKeyFileException;
import com.kunzisoft.keepass.database.exception.KeyFileEmptyException;
import com.kunzisoft.keepass.utils.MemUtil;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.UnsupportedEncodingException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.UUID;

public abstract class PwDatabase<Group extends PwGroup, Entry extends PwEntry> {

    public static final UUID UUID_ZERO = new UUID(0,0);

    // Algorithm used to encrypt the database
    protected PwEncryptionAlgorithm algorithm;

    protected byte masterKey[] = new byte[32];
    protected byte[] finalKey;

    protected PwIconFactory iconFactory = new PwIconFactory();

    protected LinkedHashMap<PwNodeId, Group> groupIndexes = new LinkedHashMap<>();
    protected LinkedHashMap<PwNodeId, Entry> entryIndexes = new LinkedHashMap<>();

    public abstract PwVersion getVersion();

    public PwIconFactory getIconFactory() {
        return iconFactory;
    }

    public byte[] getMasterKey() {
        return masterKey;
    }

    public void setMasterKey(byte[] masterKey) {
        this.masterKey = masterKey;
    }

    public byte[] getFinalKey() {
        return finalKey;
    }

    protected abstract byte[] getMasterKey(String key, InputStream keyInputStream)
			throws InvalidKeyFileException, IOException;

    public void retrieveMasterKey(String key, InputStream keyInputStream)
            throws InvalidKeyFileException, IOException {
                masterKey = getMasterKey(key, keyInputStream);
            }

    protected byte[] getCompositeKey(String key, InputStream keyInputStream)
            throws InvalidKeyFileException, IOException {
                assert(key != null && keyInputStream != null);

                byte[] fileKey = getFileKey(keyInputStream);

                byte[] passwordKey = getPasswordKey(key);

                MessageDigest md;
                try {
                    md = MessageDigest.getInstance("SHA-256");
                } catch (NoSuchAlgorithmException e) {
                    throw new IOException("SHA-256 not supported");
                }

                md.update(passwordKey);

                return md.digest(fileKey);
    }

    protected byte[] getFileKey(InputStream keyInputStream)
            throws InvalidKeyFileException, IOException {
                assert(keyInputStream != null);

                ByteArrayOutputStream bos = new ByteArrayOutputStream();
                MemUtil.copyStream(keyInputStream, bos);
                byte[] keyData = bos.toByteArray();

                ByteArrayInputStream bis = new ByteArrayInputStream(keyData);
                byte[] key = loadXmlKeyFile(bis);
                if ( key != null ) {
                    return key;
                }

                long fileSize = keyData.length;
                if ( fileSize == 0 ) {
                    throw new KeyFileEmptyException();
                } else if ( fileSize == 32 ) {
                    return keyData;
                } else if ( fileSize == 64 ) {
                    byte[] hex = new byte[64];

                    try {
                        return hexStringToByteArray(new String(keyData));
                    } catch (IndexOutOfBoundsException e) {
                        // Key is not base 64, treat it as binary data
                    }
                }

                MessageDigest md;
                try {
                    md = MessageDigest.getInstance("SHA-256");
                } catch (NoSuchAlgorithmException e) {
                    throw new IOException("SHA-256 not supported");
                }
                //SHA256Digest md = new SHA256Digest();
                byte[] buffer = new byte[2048];
                int offset = 0;

                try {
                    md.update(keyData);
                } catch (Exception e) {
                    System.out.println(e.toString());
                }

                return md.digest();
            }

    protected abstract byte[] loadXmlKeyFile(InputStream keyInputStream);

    public static byte[] hexStringToByteArray(String s) {
        int len = s.length();
        byte[] data = new byte[len / 2];
        for (int i = 0; i < len; i += 2) {
            data[i / 2] = (byte) ((Character.digit(s.charAt(i), 16) << 4)
                                 + Character.digit(s.charAt(i+1), 16));
        }
        return data;
    }

    public boolean validatePasswordEncoding(String key) {
        if (key == null)
            return false;

        String encoding = getPasswordEncoding();

        byte[] bKey;
        try {
            bKey = key.getBytes(encoding);
        } catch (UnsupportedEncodingException e) {
            return false;
        }

        String reencoded;

        try {
            reencoded = new String(bKey, encoding);
        } catch (UnsupportedEncodingException e) {
            return false;
        }

        return key.equals(reencoded);
    }

    protected abstract String getPasswordEncoding();

    public byte[] getPasswordKey(String key) throws IOException {
        if ( key == null)
            throw new IllegalArgumentException( "Key cannot be empty." ); // TODO

        MessageDigest md;
        try {
            md = MessageDigest.getInstance("SHA-256");
        } catch (NoSuchAlgorithmException e) {
            throw new IOException("SHA-256 not supported");
        }

        byte[] bKey;
        try {
            bKey = key.getBytes(getPasswordEncoding());
        } catch (UnsupportedEncodingException e) {
            assert false;
            bKey = key.getBytes();
        }
        md.update(bKey, 0, bKey.length );

        return md.digest();
    }

    public abstract long getNumberKeyEncryptionRounds();

    public abstract void setNumberKeyEncryptionRounds(long rounds) throws NumberFormatException;

    public PwEncryptionAlgorithm getEncryptionAlgorithm() {
        if (algorithm != null)
            return algorithm;
        return PwEncryptionAlgorithm.AES_Rijndael;
    }

    public void setEncryptionAlgorithm(PwEncryptionAlgorithm algorithm) {
        this.algorithm = algorithm;
    }

    public abstract List<PwEncryptionAlgorithm> getAvailableEncryptionAlgorithms();

    /*
     * -------------------------------------
     *          Node Creation
     * -------------------------------------
     */

    public abstract void populateNodesIndexes();

    public abstract PwNodeId newGroupId();

    public abstract PwNodeId newEntryId();

    public abstract Group createGroup();

    public abstract Group getRootGroup();

    /*
     * -------------------------------------
     *          Index Manipulation
     * -------------------------------------
     */

    /**
     * Determine if an id number is already in use
     *
     * @param id
     *            ID number to check for
     * @return True if the ID is used, false otherwise
     */
    public boolean isGroupIdUsed(PwNodeId id) {
        return groupIndexes.containsKey(id);
    }

    public Collection<Group> getGroupIndexes() {
        return groupIndexes.values();
    }

    public void setGroupIndexes(List<Group> groupList) {
        this.groupIndexes.clear();
        for (Group currentGroup : groupList) {
            this.groupIndexes.put(currentGroup.getNodeId(), currentGroup);
        }
    }

    public Group getGroupById(PwNodeId id) {
        return this.groupIndexes.get(id);
    }

    public void addGroupIndex(Group group) {
        this.groupIndexes.put(group.getNodeId(), group);
    }

    public void removeGroupIndex(Group group) {
		this.groupIndexes.remove(group.getNodeId());
	}

    public int numberOfGroups() {
        return groupIndexes.size();
    }

    public boolean isEntryIdUsed(PwNodeId id) {
        return entryIndexes.containsKey(id);
    }

    public Collection<Entry> getEntryIndexes() {
        return entryIndexes.values();
    }

    public Entry getEntryById(PwNodeId id) {
        return this.entryIndexes.get(id);
    }

    public void addEntryIndex(Entry entry) {
        this.entryIndexes.put(entry.getNodeId(), entry);
    }

    public void removeEntryIndex(Entry entry) {
		this.entryIndexes.remove(entry.getNodeId());
	}

    public int numberOfEntries() {
        return entryIndexes.size();
    }

    /*
     * -------------------------------------
     *          Node Manipulation
     * -------------------------------------
     */

    protected void addGroupTo(Group newGroup, @Nullable Group parent) {
		// Add tree to parent tree
		if (parent != null)
        	parent.addChildGroup(newGroup);
        newGroup.setParent(parent);
        addGroupIndex(newGroup);
    }

    protected void removeGroupFrom(Group groupToRemove, Group parent) {
        // Remove tree from parent tree
        if (parent != null) {
            parent.removeChildGroup(groupToRemove);
        }
		removeGroupIndex(groupToRemove);
    }

    protected void addEntryTo(Entry newEntry, @Nullable Group parent) {
        // Add entry to parent
		if (parent != null)
			parent.addChildEntry(newEntry);
        newEntry.setParent(parent);
        addEntryIndex(newEntry);
    }

    protected void removeEntryFrom(Entry entryToRemove, Group parent) {
        // Remove entry for parent
        if (parent != null) {
            parent.removeChildEntry(entryToRemove);
        }
        removeEntryIndex(entryToRemove);
    }

    // TODO Delete group
    public void undoDeleteGroupFrom(Group group, Group origParent) {
        addGroupTo(group, origParent);
    }

    public void undoDeleteEntryFrom(Entry entry, Group origParent) {
        addEntryTo(entry, origParent);
    }

    public abstract boolean isBackup(Group group);

    public boolean isGroupSearchable(Group group, boolean omitBackup) {
        return group != null;
    }
}

File: app/src/main/java/com/kunzisoft/keepass/database/element/PwDatabase.kt
/*
 * Copyright 2019 Jeremy Jamet / Kunzisoft.
 *     
 * This file is part of KeePass DX.
 *
 *  KeePass DX is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  KeePass DX is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with KeePass DX.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.kunzisoft.keepass.database.element

import com.kunzisoft.keepass.database.exception.InvalidKeyFileException
import com.kunzisoft.keepass.database.exception.KeyFileEmptyException
import com.kunzisoft.keepass.utils.MemUtil

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.InputStream
import java.io.UnsupportedEncodingException
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.util.LinkedHashMap
import java.util.UUID

abstract class PwDatabase<Group : PwGroup<*, Group, Entry>, Entry : PwEntry<Group, Entry>> {

    // Algorithm used to encrypt the database
    protected var algorithm: PwEncryptionAlgorithm? = null

    var masterKey = ByteArray(32)
    var finalKey: ByteArray? = null
        protected set

    var iconFactory = PwIconFactory()
        protected set

    private var groupIndexes = LinkedHashMap<PwNodeId<*>, Group>()
    private var entryIndexes = LinkedHashMap<PwNodeId<*>, Entry>()

    abstract val version: String

    protected abstract val passwordEncoding: String

    abstract var numberKeyEncryptionRounds: Long

    var encryptionAlgorithm: PwEncryptionAlgorithm
        get() {
            return algorithm ?: PwEncryptionAlgorithm.AESRijndael
        }
        set(algorithm) {
            this.algorithm = algorithm
        }

    abstract val availableEncryptionAlgorithms: List<PwEncryptionAlgorithm>

    var rootGroup: Group? = null

    @Throws(InvalidKeyFileException::class, IOException::class)
    protected abstract fun getMasterKey(key: String?, keyInputStream: InputStream?): ByteArray

    @Throws(InvalidKeyFileException::class, IOException::class)
    fun retrieveMasterKey(key: String?, keyInputStream: InputStream?) {
        masterKey = getMasterKey(key, keyInputStream)
    }

    @Throws(InvalidKeyFileException::class, IOException::class)
    protected fun getCompositeKey(key: String, keyInputStream: InputStream): ByteArray {
        val fileKey = getFileKey(keyInputStream)

        val passwordKey = getPasswordKey(key)

        val md: MessageDigest
        try {
            md = MessageDigest.getInstance("SHA-256")
        } catch (e: NoSuchAlgorithmException) {
            throw IOException("SHA-256 not supported")
        }

        md.update(passwordKey)

        return md.digest(fileKey)
    }

    @Throws(InvalidKeyFileException::class, IOException::class)
    protected fun getFileKey(keyInputStream: InputStream): ByteArray {

        val bos = ByteArrayOutputStream()
        MemUtil.copyStream(keyInputStream, bos)
        val keyData = bos.toByteArray()

        val bis = ByteArrayInputStream(keyData)
        val key = loadXmlKeyFile(bis)
        if (key != null) {
            return key
        }

        val fileSize = keyData.size.toLong()
        if (fileSize == 0L) {
            throw KeyFileEmptyException()
        } else if (fileSize == 32L) {
            return keyData
        } else if (fileSize == 64L) {
            try {
                return hexStringToByteArray(String(keyData))
            } catch (e: IndexOutOfBoundsException) {
                // Key is not base 64, treat it as binary data
            }
        }

        val md: MessageDigest
        try {
            md = MessageDigest.getInstance("SHA-256")
        } catch (e: NoSuchAlgorithmException) {
            throw IOException("SHA-256 not supported")
        }

        try {
            md.update(keyData)
        } catch (e: Exception) {
            println(e.toString())
        }

        return md.digest()
    }

    protected abstract fun loadXmlKeyFile(keyInputStream: InputStream): ByteArray?

    open fun validatePasswordEncoding(key: String?): Boolean {
        if (key == null)
            return false

        val encoding = passwordEncoding

        val bKey: ByteArray
        try {
            bKey = key.toByteArray(charset(encoding))
        } catch (e: UnsupportedEncodingException) {
            return false
        }

        val reEncoded: String
        try {
            reEncoded = String(bKey, charset(encoding))
        } catch (e: UnsupportedEncodingException) {
            return false
        }
        return key == reEncoded
    }

    @Throws(IOException::class)
    fun getPasswordKey(key: String?): ByteArray {
        if (key == null)
            throw IllegalArgumentException("Key cannot be empty.") // TODO

        val md: MessageDigest
        try {
            md = MessageDigest.getInstance("SHA-256")
        } catch (e: NoSuchAlgorithmException) {
            throw IOException("SHA-256 not supported")
        }

        val bKey: ByteArray = try {
            key.toByteArray(charset(passwordEncoding))
        } catch (e: UnsupportedEncodingException) {
            key.toByteArray()
        }

        md.update(bKey, 0, bKey.size)

        return md.digest()
    }

    /*
     * -------------------------------------
     *          Node Creation
     * -------------------------------------
     */

    abstract fun newGroupId(): PwNodeId<*>

    abstract fun newEntryId(): PwNodeId<*>

    abstract fun createGroup(): Group

    abstract fun createEntry(): Entry

    /*
     * -------------------------------------
     *          Index Manipulation
     * -------------------------------------
     */

    fun doForEachGroupInIndex(action: (Group) -> Unit) {
        for (group in groupIndexes) {
            action.invoke(group.value)
        }
    }

    /**
     * Determine if an id number is already in use
     *
     * @param id
     * ID number to check for
     * @return True if the ID is used, false otherwise
     */
    fun isGroupIdUsed(id: PwNodeId<*>): Boolean {
        return groupIndexes.containsKey(id)
    }

    fun getGroupIndexes(): Collection<Group> {
        return groupIndexes.values
    }

    fun setGroupIndexes(groupList: List<Group>) {
        this.groupIndexes.clear()
        for (currentGroup in groupList) {
            this.groupIndexes[currentGroup.nodeId] = currentGroup
        }
    }

    fun getGroupById(id: PwNodeId<*>): Group? {
        return this.groupIndexes[id]
    }

    fun addGroupIndex(group: Group) {
        this.groupIndexes[group.nodeId] = group
    }

    fun removeGroupIndex(group: Group) {
        this.groupIndexes.remove(group.nodeId)
    }

    fun numberOfGroups(): Int {
        return groupIndexes.size
    }

    fun doForEachEntryInIndex(action: (Entry) -> Unit) {
        for (entry in entryIndexes) {
            action.invoke(entry.value)
        }
    }

    fun isEntryIdUsed(id: PwNodeId<*>): Boolean {
        return entryIndexes.containsKey(id)
    }

    fun getEntryIndexes(): Collection<Entry> {
        return entryIndexes.values
    }

    fun getEntryById(id: PwNodeId<*>): Entry? {
        return this.entryIndexes[id]
    }

    fun addEntryIndex(entry: Entry) {
        this.entryIndexes[entry.nodeId] = entry
    }

    fun removeEntryIndex(entry: Entry) {
        this.entryIndexes.remove(entry.nodeId)
    }

    fun numberOfEntries(): Int {
        return entryIndexes.size
    }

    open fun clearCache() {
        this.groupIndexes.clear()
        this.entryIndexes.clear()
    }

    /*
     * -------------------------------------
     *          Node Manipulation
     * -------------------------------------
     */

    fun addGroupTo(newGroup: Group, parent: Group?) {
        // Add tree to parent tree
        parent?.addChildGroup(newGroup)
        newGroup.parent = parent
        addGroupIndex(newGroup)
    }

    fun removeGroupFrom(groupToRemove: Group, parent: Group?) {
        // Remove tree from parent tree
        parent?.removeChildGroup(groupToRemove)
        removeGroupIndex(groupToRemove)
    }

    fun addEntryTo(newEntry: Entry, parent: Group?) {
        // Add entry to parent
        parent?.addChildEntry(newEntry)
        newEntry.parent = parent
        addEntryIndex(newEntry)
    }

    open fun removeEntryFrom(entryToRemove: Entry, parent: Group?) {
        // Remove entry for parent
        parent?.removeChildEntry(entryToRemove)
        removeEntryIndex(entryToRemove)
    }

    // TODO Delete group
    fun undoDeleteGroupFrom(group: Group, origParent: Group?) {
        addGroupTo(group, origParent)
    }

    open fun undoDeleteEntryFrom(entry: Entry, origParent: Group?) {
        addEntryTo(entry, origParent)
    }

    abstract fun isBackup(group: Group): Boolean

    fun isGroupSearchable(group: Group?, omitBackup: Boolean): Boolean {
        if (group == null)
            return false
        if (omitBackup && isBackup(group))
            return false
        return true
    }

    companion object {

        val UUID_ZERO = UUID(0, 0)

        fun hexStringToByteArray(s: String): ByteArray {
            val len = s.length
            val data = ByteArray(len / 2)
            var i = 0
            while (i < len) {
                data[i / 2] = ((Character.digit(s[i], 16) shl 4) + Character.digit(s[i + 1], 16)).toByte()
                i += 2
            }
            return data
        }
    }
}
